replica/distributed_loader.cc

```diff
@@ -401,9 +401,38 @@ distributed_loader::make_sstables_available(sstables::sstable_directory& dir, sh
     co_return new_sstables.size();
 }
 
-sstables::shared_sstable make_sstable(replica::table& table, fs::path dir, int64_t generation_value) {
-    auto& sstm = table.get_sstables_manager();
-    return sstm.make_sstable(table.schema(), dir.native(), sstables::generation_from_value(generation_value), sstm.get_highest_supported_format(), sstables::sstable_format_types::big, gc_clock::now(), &error_handler_gen_for_upload_dir);
+namespace {
+
+std::unique_ptr<sstables::generation_generator>
+create_generation_generator(sstables::generation_type highest_generation) {
+    auto generation_base = highest_generation.value() / smp::count + 1;
+    return std::make_unique<sstables::int_generation_generator>(
+        generation_base * smp::count + this_shard_id());
+}
+
+class sstable_creator {
+    replica::table& _table;
+    fs::path _dir;
+    std::unique_ptr<sstables::generation_generator> _generation_generator;
+public:
+    sstable_creator(replica::table& table,
+                    fs::path dir,
+                    std::unique_ptr<sstables::generation_generator> generation_generator)
+        : _table{table}
+        , _dir{dir}
+        , _generation_generator{std::move(generation_generator)} {}
+    sstables::shared_sstable operator()() {
+        auto& sstm = _table.get_sstables_manager();
+        return sstm.make_sstable(_table.schema(),
+                                 _dir.native(),
+                                 _generation_generator->generate(),
+                                 sstm.get_highest_supported_format(),
+                                 sstables::sstable_format_types::big,
+                                 gc_clock::now(),
+                                 &error_handler_gen_for_upload_dir);
+    }
+};
```

Comment:
Can't we fold this into the `sstable_manager` instead? It is already mandatory to create sstables.

Reply:
good point. will implement this and fix the test failures in the next revision !

Reply:
`sstable_manager` does not track the latest generation so far. and there are multiple different use cases where the generation is passed to `make_sstable()`. i am not sure how to consolidate them without sacrificing the flexibility. could you be more specific?

Reply:
I just dislike creating a brand new sharded service just for generating generations. Its one more service we have to propagate down to every place which wants to create sstables. As the sstables-manager is already mandatory on all these sites, I though it could also take on the role of generating the generations. I don't have anything more specific in mind than that, sorry.

Reply:
i see. and i concur with you. the sharded service is overkill and does not fit in this use case. as it is not symmetric. and the states are shared across multiple shards. the latest revision of this PR just uses a degraded form of generation identifier generator.

Reply:
Very good idea on moving it to sstable manager. replica::table talks to sstable manager for everything, including picking a new sstable with a new generation.

today, we have:

sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_sstable(_schema, dir, calculate_generation_for_new_table(), sstm.get_highest_supported_format(), sstables::sstable::format_types::big);
}

but we can change it to become:

sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_new_sstable(_schema, dir);
}

Reply:
i think the change you are proposing implies that we can move `calculate_generation_for_new_table()` into `sstables_manager`, but we only hsave two sstables_managers in each scylla server, one for user tables, another for system tables. so, this in turn has an assumption that the highest generation id can be shared across different tables, and it should be stored in `sstables_manager`. but currently, each table keeps track of its own generation id, and it uses it to tell if it is ready for write. so, i am afraid that we cannot store the highest generation id in `sstables_manager`, neither can we use `sstables_manager` for creating new generation id.

one of the goals of generation generator design is to avoid segmenting the implementations of how we generation id generation, and to consolidate them to a single place. so, i think we might need to keep 
```c++
sstables::shared_sstable table::make_sstable(sstring dir) {
    auto& sstm = get_sstables_manager();
    return sstm.make_sstable(_schema, dir, 
        calculate_generation_for_new_table(),  
        sstm.get_highest_supported_format(),
        sstables::sstable::format_types::big);
}
```

as it is.